niworkflows/utils/spaces.py

```diff
@@ -0,0 +1,418 @@
+    >>> Space(name='fsaverage6', spec={'den': '10k'})
+    Space(name='fsaverage', cohort=None, spec={'den': '41k'})
+
+    >>> Space('func').standard
+    False
+
+    >>> Space('MNI152Lin').standard
+    True
+
+    >>> Space('MNIPediatricAsym', {'cohort': 1}).standard
+    True
+
+    >>> Space('func') == Space('func')
+    True
+
+    >>> Space('func') != Space('MNI152Lin')
+    True
+
+    >>> Space('MNI152Lin', {'res': 1}) == Space('MNI152Lin', {'res': 1})
+    True
+
+    >>> Space('MNI152Lin', {'res': 1}) == Space('MNI152Lin', {'res': 2})
+    False
+
+    >>> sp1 = Space('MNIPediatricAsym', {'cohort': 1})
+    >>> sp2 = Space('MNIPediatricAsym', {'cohort': 2})
+    >>> sp1 == sp2
+    False
+
+    >>> sp1 = Space('MNIPediatricAsym', {'res': 1, 'cohort': 1})
+    >>> sp2 = Space('MNIPediatricAsym', {'cohort': 1, 'res': 1})
+    >>> sp1 == sp2
+    True
+
+    >>> Space.from_string("MNI152NLin2009cAsym")
+    [Space(name='MNI152NLin2009cAsym', cohort=None, spec={})]
+
+    >>> # Bad name
+    >>> Space.from_string("shouldraise")
+    Traceback (most recent call last):
+      ...
+    ValueError: space identifier "shouldraise" is invalid.
+    ...
+
+    >>> # Missing cohort
+    >>> Space.from_string("MNIPediatricAsym")
+    Traceback (most recent call last):
+      ...
+    ValueError: standard space "MNIPediatricAsym" is not fully defined.
+    ...
+
+    >>> Space.from_string("MNIPediatricAsym:cohort-1")
+    [Space(name='MNIPediatricAsym', cohort='1', spec={})]
+
+    >>> Space.from_string("MNIPediatricAsym:cohort-1:cohort-2")
+    [Space(name='MNIPediatricAsym', cohort='1', spec={}),
+     Space(name='MNIPediatricAsym', cohort='2', spec={})]
+
+    >>> Space.from_string("MNIPediatricAsym:cohort-5:cohort-6:res-2")
+    [Space(name='MNIPediatricAsym', cohort='5', spec={'res': '2'}),
+     Space(name='MNIPediatricAsym', cohort='6', spec={'res': '2'})]
+
+    >>> Space.from_string("MNIPediatricAsym:cohort-5:cohort-6:res-2:res-iso1.6mm")
+    [Space(name='MNIPediatricAsym', cohort='5', spec={'res': '2'}),
+     Space(name='MNIPediatricAsym', cohort='5', spec={'res': 'iso1.6mm'}),
+     Space(name='MNIPediatricAsym', cohort='6', spec={'res': '2'}),
+     Space(name='MNIPediatricAsym', cohort='6', spec={'res': 'iso1.6mm'})]
+
+    """
+
+    _standard_spaces = tuple(_tfapi.templates())
+
+    name: str = attr.ib(default=None)
+    """Unique name designating this space."""
+    cohort: str = attr.ib(init=False, default=None)
+    """An attribute to accomodate cohorts from TemplateFlow."""
+    spec: typing.Dict = attr.ib(factory=dict)
+    """The dictionary of specs."""
+    standard: bool = attr.ib(default=False, repr=False)
+    """Whether this space is standard or not."""
+    dim: int = attr.ib(default=3, repr=False)
+    """Dimensionality of the sampling manifold."""
+
+    def __attrs_post_init__(self):
+        """Extract cohort out of spec."""
+        if self.spec is None:
+            object.__setattr__(self, "spec", {})
```

Comment:
This trick is necessary when the class is frozen.

Reply:
Why are we freezing it? It seems if we need to hack around, then we're probably not using the API as intended.

Reply:
Once you create a Space object, it should remain untouched. The trick is suggested within the attrs documentation for this particular situation.